File: objc/Project2/Project2Tests/LifeCycleTests.m
//
//  LifeCycleTests.m
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

#import <XCTest/XCTest.h>

@interface LifeCycleTests : XCTestCase

@end

@implementation LifeCycleTests

+ (void)setUp {
    NSLog(@"[Static] setUp");
}

- (void)setUp {
    NSLog(@"setUp");
}

- (BOOL)setUpWithError:(NSError *__autoreleasing  _Nullable *)error {
    NSLog(@"setUpWithError");
    *error = [[NSError alloc] initWithDomain:@"TestError" code:1003 userInfo:@{NSLocalizedDescriptionKey: @"Localized Test Error!"}];
    return NO; // NO이면 테스트 안함 - tearDown Life Cycle은 그대로 돌아감. Error가 발생할 때도 마찬가지
}

//

+ (void)tearDown {
    NSLog(@"[Static] tearDown");
}

- (void)tearDown {
    NSLog(@"tearDown");
}

- (BOOL)tearDownWithError:(NSError *__autoreleasing  _Nullable *)error {
    NSLog(@"tearDownWithError");
    return NO; // NO여도 + (void)tearDown은 돌아감. Error가 발생할 때도 마찬가지
}

//

- (void)testExample {
    NSLog(@"testExample");
}

@end

File: swift/Project2/Project2Tests/LifecycleTests.swift
//
//  LifecycleTests.swift
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

import XCTest

// Error가 발생해도 tearDown Life Cycle은 그대로 돌아감 - test는 안 돌아감
class LifecycleTests: XCTestCase {
    override class func setUp() {
        print(#function)
    }
    
    override class func tearDown() {
        print(#function)
    }
    
    override func setUp() {
        print(#function)
    }
    
    override func tearDown() {
        print(#function)
    }
    
    override func setUpWithError() throws {
        print(#function)
    }
    
    override func tearDownWithError() throws {
        print(#function)
    }
    
    func testExample() {
        print(#function)
        
        addTeardownBlock {
            print("In first tearDown block.")
        }
        
        print(#function)
    }
}
